File: psets/pset2/generate_g.jl
#!/usr/bin/julia -f

function generate_g!(G)
    n = size(G, 1)
    @inbounds for i in 1:(n - 1)
        # Auto vectorization already works here...
        @simd for j in 1:n
            G[j, i] = ifelse(j < i, 0, ifelse(j == i, 1, -1))
        end
    end
    # Auto vectorization already works here...
    @inbounds @simd for j in 1:n
        G[j, n] = 1
    end
    G
end

generate_g_c!(G) =
    ccall((:generate_g, "./libgen_g.so"),
          Void, (Ptr{Void}, Csize_t), G, size(G, 1))

generate_g_cilk!(G) =
    ccall((:generate_g_cilk, "./libgen_g.so"),
          Void, (Ptr{Void}, Csize_t), G, size(G, 1))

generate_g_omp!(G) =
    ccall((:generate_g_omp, "./libgen_g.so"),
          Void, (Ptr{Void}, Csize_t), G, size(G, 1))

fill_memset(G) =
    ccall(:memset, Ptr{Void}, (Ptr{Void}, Cint, Csize_t), G, 0, sizeof(G))

fill_cilk(G) =
    ccall((:fill_cilk, "./libgen_g.so"),
          Void, (Ptr{Void}, Csize_t), G, length(G))

fill_omp(G) =
    ccall((:fill_omp, "./libgen_g.so"),
          Void, (Ptr{Void}, Csize_t), G, length(G))

function generate_g(n)
    G = Array{Float64}(n, n)
    generate_g!(G)
end

File: psets/pset2/generate_g.c
// compile using GCC 5.2.0 with the following command line
// gcc -fopenmp -fcilkplus -mavx2 -fPIC -Ofast generate_g.c -o libgen_g.so
//     -shared -Wall -Wextra

#include <stddef.h>
#include <cilk/cilk.h>

void
fill_cilk(double *__restrict__ ary, size_t size)
{
    cilk_for (size_t i = 0;i < size;i++) {
        ary[i] = 0;
    }
}

void
fill_omp(double *__restrict__ ary, size_t size)
{
#pragma omp parallel for
    for (size_t i = 0;i < size;i++) {
        ary[i] = 0;
    }
}

void
generate_g(double *__restrict__ ptr, size_t n)
{
    for (size_t i = 0;i < n - 1;i++) {
        double *ptr2 = ptr + i * n;
#pragma simd
        for (size_t j = 0;j < n;j++) {
            ptr2[j] = j < i ? 0 : (j == i ? 1 : -1);
        }
    }
#pragma simd
    for (size_t j = 0;j < n;j++) {
        ptr[(n - 1) * n + j] = 1;
    }
}

void
generate_g_cilk(double *__restrict__ ptr, size_t n)
{
    cilk_for (size_t i = 0;i < n - 1;i++) {
        double *ptr2 = ptr + i * n;
#pragma simd
        for (size_t j = 0;j < n;j++) {
            ptr2[j] = j < i ? 0 : (j == i ? 1 : -1);
        }
    }
    cilk_for (size_t j = 0;j < n;j++) {
        ptr[(n - 1) * n + j] = 1;
    }
}

void
generate_g_omp(double *__restrict__ ptr, size_t n)
{
#pragma omp parallel for
    for (size_t i = 0;i < n - 1;i++) {
        double *ptr2 = ptr + i * n;
#pragma simd
        for (size_t j = 0;j < n;j++) {
            ptr2[j] = j < i ? 0 : (j == i ? 1 : -1);
        }
    }
#pragma omp parallel for
    for (size_t j = 0;j < n;j++) {
        ptr[(n - 1) * n + j] = 1;
    }
}
